### PR TITLE
WFLY-9547, WFLY-9548 Clustering API enhancements

### PR DESCRIPTION
**Pull request by pferraro**
https://issues.jboss.org/browse/WFLY-9547
https://issues.jboss.org/browse/WFLY-9548

This PR addresses transient failures in the clustering testsuite as seen in:
https://ci.wildfly.org/viewLog.html?buildId=81831&buildTypeId=WF_PullRequest_Windows

### CODE REVIEW

File: clustering/server/src/main/java/org/wildfly/clustering/server/dispatcher/ChannelCommandDispatcherFactoryBuilderProvider.java [hunk 1]
```diff
@@ -21,12 +21,14 @@
...
 package org.wildfly.clustering.server.dispatcher;
 
+import org.kohsuke.MetaInfServices;
 import org.wildfly.clustering.spi.DistributedGroupBuilderProvider;
 
 /**
  * Provides the requisite builders for creating a channel-based {@link org.wildfly.clustering.dispatcher.CommandDispatcherFactory}.
  * @author Paul Ferraro
  */
+@MetaInfServices(org.wildfly.clustering.spi.DistributedGroupBuilderProvider.class)
```
Comment (rhusar) [inline clustering/server/src/main/java/org/wildfly/clustering/server/dispatcher/ChannelCommandDispatcherFactoryBuilderProvider.java]: No need for FQCN.

File: clustering/server/src/main/java/org/wildfly/clustering/server/dispatcher/CommandDispatcherFactoryAliasBuilderProvider.java [hunk 2]
```diff
@@ -22,13 +22,15 @@
...
 
+import org.kohsuke.MetaInfServices;
 import org.wildfly.clustering.server.GroupJndiNameFactory;
 import org.wildfly.clustering.server.GroupRequirementAliasBuilderProvider;
 import org.wildfly.clustering.spi.ClusteringRequirement;
 
 /**
  * @author Paul Ferraro
  */
+@MetaInfServices(org.wildfly.clustering.spi.GroupAliasBuilderProvider.class)
```
Comment (rhusar) [inline clustering/server/src/main/java/org/wildfly/clustering/server/dispatcher/CommandDispatcherFactoryAliasBuilderProvider.java]: Could use an import... and in other ~5 places. Or is that intentional?

File: clustering/api/src/main/java/org/wildfly/clustering/group/Group.java [hunk 3]
```diff
@@ -105,8 +105,15 @@ default void membershipChanged(Membership previousMembership, Membership members
     Membership getMembership();
 
     /**
-     * Indicates whether this is a local group.  A local group only ever contains a single member.
-     * @return true, if this is a local group, false otherwise.
+     * @deprecated Replaced by {@link #isSingleton()}.
```
Comment (rhusar) [inline clustering/api/src/main/java/org/wildfly/clustering/group/Group.java]: Generally speaking, I think in most cases the original Javadoc should be kept in place when deprecating -- though in this case (and seems like in all others) the updates are direct replacements, so its fine I think.

**Review by rhusar — APPROVED** [review]
👍  Great improvements; especially `Membership` fixes around previously racy API usage.

**Comment by bstansberry** [comment]
Is this a change to a public-to-an-end-user API, aka a Feature?

**Comment by pferraro** [comment]
@bstansberry It's a change to public API, but there's no feature being added.  All changes are binary/source compatible with previous releases.
This stuff is already fully covered by our existing clustering API tests:
https://github.com/wildfly/wildfly/tree/master/testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/dispatcher
https://github.com/wildfly/wildfly/tree/master/testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/provider
https://github.com/wildfly/wildfly/tree/master/testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/registry
https://github.com/wildfly/wildfly/tree/master/testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/local/dispatcher
https://github.com/wildfly/wildfly/tree/master/testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/local/provider
https://github.com/wildfly/wildfly/tree/master/testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/local/registry

**Review by bstansberry — APPROVED** [review]

**Comment by bstansberry** [comment]
@pferraro @rhusar When I put this and the CommandRPC one on master-ignore there was a failure: 

https://ci.wildfly.org/viewLog.html?buildId=82256&tab=buildResultsDiv&buildTypeId=WF_MasterIgnoreLinux

That test has been failing off and on for the past week so I'll merge this anyway, but that one should be sorted. 8 failures in the last week; maybe more if I checked the CI jobs for WFCORE.

**Comment by rhusar** [comment]
@bstansberry Thanks, we are aware of the issue and working on it. If I could I would mute the test until we figure it out.

**Pull request merged**